lib/metrics.js

```diff
@@ -0,0 +1,848 @@
+   */
+
+  /**
+   * Duration of timer in milliseconds. If undefined, timer is still running.
+   * @property {number|undefined} duration
+   */
+
+  /**
+   * Construct a new timer.
+   * @param {Metric} metric The metric to report to.
+   * @param {object} [meta] Additional metadata to include with the report.
+   */
+  constructor(metric, meta) {
+    super(metric);
+    if (meta !== undefined && typeof meta !== 'object') {
+      throw new ERR_INVALID_ARG_TYPE('meta', ['object', 'undefined'], meta);
+    }
+    this.#meta = meta;
+
+    this.start = performance.now();
+    this.end = undefined;
+    this.duration = undefined;
+  }
+
+  /**
+   * Additional metadata to include with the report.
+   * @property {object} meta
+   */
+  get meta() {
+    return mixMeta(this.metric.meta, this.#meta);
+  }
+
+  /**
+   * Stop the timer and report the duration.
+   * @param {object} [meta] Additional metadata to include with the report.
+   * @returns {number} The duration in milliseconds.
+   */
+  stop(meta) {
+    if (this.end !== undefined) return;
+    if (!this.metric.shouldReport) return;
+    this.end = performance.now();
+    this.duration = this.end - this.start;
+    this.reset(this.duration, mixMeta(this.#meta, meta));
+    return this.duration;
+  }
+
+  /**
+   * Support `using` syntax to automatically stop the timer when done.
+   */
+  [SymbolDispose]() {
+    this.stop();
+  }
+}
+
+/**
+ * A count of the number of unique values that have been seen.
+ */
+class UniqueSet extends Gauge {
+  // Use a weak set to track unique values without retaining references.
+  #seen = new SafeWeakSet();
+
+  /**
+   * The number of unique values seen.
+   * @property {number} count
+   */
+  count = 0;
+
+  /**
+   * Add value to set. If value was not already present, report it.
+   * @param {any} value The value to track in the set.
+   * @param {object} meta Additional metadata to include with the report.
+   */
+  add(value, meta) {
+    // If already seen, do nothing.
+    if (this.#seen.has(value)) {
+      return;
+    }
+
+    this.count += 1;
+    this.#seen.add(value);
+    this.reset(this.count, meta);
+  }
+}
+
+/**
+ * A gauge which periodically updates its value by calling a function and
+ * setting the value to the result.
+ */
+class PeriodicGauge extends Gauge {
+  #timer;
+  #interval;
+  #fn;
+
+  /**
+   * Construct a new periodic gauge.
+   * @param {Metric} metric The metric to report to.
+   * @param {number} interval The interval in milliseconds to update the gauge.
+   * @param {Function} fn The function to call to update the gauge.
+   */
+  constructor(metric, interval, fn) {
+    super(metric);
+
+    if (typeof interval !== 'number' || interval <= 0) {
+      throw new ERR_INVALID_ARG_TYPE('interval', ['number'], interval);
+    }
+    if (typeof fn !== 'function') {
+      throw new ERR_INVALID_ARG_TYPE('fn', ['function'], fn);
+    }
+
+    this.#timer = undefined;
+    this.#interval = interval;
+    this.#fn = fn;
```

Comment:
It would be nice to have an option to perform an initial sample of `fn` on construction

Reply:
Probably should be the default, actually. Not sure if one would ever _not_ want an initial sample though, maybe it doesn't even need an option? 🤔